File: smartadapter/src/main/java/smartadapter/listener/OnViewEventListener.java
package smartadapter.listener;

/*
 * Created by Manne Öhlund on 02/04/17.
 * Copyright © 2017 All rights reserved.
 */

import android.view.View;

import androidx.annotation.NonNull;

import io.github.manneohlund.smartrecycleradapter.R;
import smartadapter.viewholder.SmartViewHolder;
import smartadapter.viewholder.ViewEventListenerHolder;

/**
 * Callback added in {@link smartadapter.SmartRecyclerAdapter} for view events listening in {@link SmartViewHolder} extensions.
 */
public interface OnViewEventListener {

    /**
     * Default implementation returns {@link SmartViewHolder} class which
     * {@link smartadapter.SmartRecyclerAdapter} will bind to all {@link SmartViewHolder} extensions.
     *
     * <p>Can be overridden to a specific target {@link SmartViewHolder} extension.</p>
     *
     * @return {@link SmartViewHolder}.class
     */
    @NonNull
    default Class<? extends SmartViewHolder> getViewHolderType() {
        return SmartViewHolder.class;
    }

    /**
     * Default implementation returns {@link R.id#undefined} which will point to the root view of the view in the view holder.
     *
     * <p>Can be overridden to target specific view id.</p>
     *
     * @return {@link R.id#undefined}
     */
    default int getViewId() {
        return R.id.undefined;
    }

    /**
     * Default implementation returns {@link R.id#undefined} and any {@link SmartViewHolder} extensions must implement {@link ViewEventListenerHolder}.
     *
     * <p>Can be overridden with predefined view action ids.</p>
     * {@link R.id#event_on_click}
     * {@link R.id#event_on_long_click}
     *
     * @return {@link R.id#undefined}
     */
    default int getViewEventId() {
        return R.id.undefined;
    }

    /**
     * Receiver of a view events.
     *
     * @param view source view which dispatched the action
     * @param actionId callback actionId defined in {@link #getViewEventId()}
     * @param position the adapter position
     */
    void onViewEvent(@NonNull View view, int actionId, int position);
}

File: smartadapter/src/main/java/smartadapter/listener/OnViewEventListener.kt
package smartadapter.listener

/*
 * Created by Manne Öhlund on 02/04/17.
 * Copyright © 2017 All rights reserved.
 */

import android.view.View
import io.github.manneohlund.smartrecycleradapter.R
import smartadapter.Position
import smartadapter.SmartViewHolderType
import smartadapter.ViewEventId
import smartadapter.ViewId
import smartadapter.viewholder.SmartViewHolder
import smartadapter.viewholder.ViewEventListenerHolder

/**
 * Callback added in [smartadapter.SmartRecyclerAdapter] for view events listening in [SmartViewHolder] extensions.
 */
@Deprecated("Will be removed soon, use extension library 'io.github.manneohlund:smart-recycler-adapter-listeners'")
interface OnViewEventListener {

    /**
     * Default implementation returns [SmartViewHolder] class which
     * [smartadapter.SmartRecyclerAdapter] will bind to all [SmartViewHolder] extensions.
     *
     * Can be overridden to a specific target [SmartViewHolder] extension.
     */
    val viewHolderType: SmartViewHolderType
        get() = SmartViewHolder::class

    /**
     * Default implementation returns [R.id.undefined] which will point to the root view of the view in the view holder.
     *
     * Can be overridden to target specific view id.
     */
    val viewId: ViewId
        get() = R.id.undefined

    /**
     * Default implementation returns [R.id.undefined] and any [SmartViewHolder] extensions must implement [ViewEventListenerHolder].
     *
     * Can be overridden with predefined view action ids.
     * [R.id.event_on_click]
     * [R.id.event_on_long_click]
     */
    val viewEventId: ViewEventId
        get() = R.id.undefined

    /**
     * Receiver of a view events.
     *
     * @param view source view which dispatched the action
     * @param viewEventId callback viewEventId
     * @param position the adapter position
     */
    fun onViewEvent(view: View, viewEventId: ViewEventId, position: Position)
}

/**
 * Helper method to provide lambda call to interface instances of [OnViewEventListener].
 */
inline fun onViewEventListener(crossinline viewEvent: (
    view: View,
    viewEventId: ViewEventId,
    position: Position) -> Unit) = object : OnViewEventListener {
    override fun onViewEvent(view: View, viewEventId: ViewEventId, position: Position) {
        viewEvent(view, viewEventId, position)
    }
}